### PR TITLE
Shore up dataflow |-> coord tail protocol by removing the redundant `Complete` variant and changing when `Dropped` is sent.

### LINKED ISSUE #9479: TailResponse: Remove Complete, Dropped states
The `TailResponse` enum currently has the following states:
* `Progress(Antichain)`
* `Rows(Vec<_>)`
* `Complete`
* `Dropped`

In Timely, an empty antichain would include the dropped/complete case. Determine if this is true for tail sinks and if yes, remove the complete and dropped variants (or make sure that the variants can be removed by adjusting the implementation).

### PR DESCRIPTION
**Pull request by umanwizard**
This PR is described in the commit messages, not in this PR description field.

Fixes #9479

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/9889)

### CODE REVIEW

**Review by aljoscha — APPROVED** [review]
I _think_ this makes sense! And I'm reasonably sure that the unit tests for tail work.

File: src/dataflow/src/sink/tail.rs [hunk 1]
```diff
@@ -135,9 +145,11 @@ fn tail<G>(
                 }
             }
 
-            // Emit data if configured, otherwise it is an error to have data to send.
-            if let Some(tail_protocol) = &mut tail_protocol {
-                tail_protocol.send_rows(results);
+            if active_worker {
+                // Emit data if configured, otherwise it is an error to have data to send.
```
Comment (antiguru) [inline src/dataflow/src/sink/tail.rs:149]: Is it possible to have `active_worker && tail_protocol_handle is None`? This seems odd to me and if it's not the case, we wouldn't actually need the `active_worker` variable, but only decide to do work based on that the handle is `Some(_)`. But, I might be missing the relevant code!

**Review by antiguru — APPROVED** [review]
LGTM, minus my comment whether `active_worker` is actually needed. Thanks a bunch!

File: src/dataflow/src/sink/tail.rs (hunk 1 above)
Comment (umanwizard) [reply src/dataflow/src/sink/tail.rs:149]: > Is it possible to have active_worker && tail_protocol_handle is None? 

Yes. After shutdown is initiated, this code runs:

```
        // Inform the coordinator that we have been dropped,
        // and destroy the tail protocol so the sink operator
        // can't send spurious messages while shutting down.
        Some(Box::new(scopeguard::guard((), move |_| {
            if let Some(tail_protocol_handle) = tail_protocol_weak.upgrade() {
                std::mem::drop(tail_protocol_handle.borrow_mut().take())
            }
        })))
```

which causes the tail protocol to be dropped (and both this scopeguard and the operator are sharing the same tail protocol object, behind an Rc<RefCell<_>>), but the dataflow operator can still be getting progress messages.

**Pull request merged**